src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/ObjectSynchronizer.java

```diff
@@ -43,8 +43,10 @@ public void update(Observable o, Object data) {
   }
 
   private static synchronized void initialize(TypeDataBase db) throws WrongTypeException {
-    Type type = db.lookupType("ObjectSynchronizer");
-    inUseList = type.getAddressField("_in_use_list").getValue();
+    Type objectSynchronizerType = db.lookupType("ObjectSynchronizer");
+    Type monitorListType = db.lookupType("MonitorList");
+    Address monitorListAddr = objectSynchronizerType.getField("_in_use_list").getStaticFieldAddress();
+    inUseListHead = monitorListType.getAddressField("_head").getAddress(monitorListAddr);
```

Comment:
It is not. It is the value of `_head` when ObjectSynchronizer is initialised. My understanding of the serviceability agent is very limited, but from what I understood the JVM does not run when we are attached. So no code should add to the list. 

If that was a requirement then both this and the previous implementation are completely broken.

Reply:
> My understanding of the serviceability agent is very limited, but from what I understood the JVM does not run when we are attached. So no code should add to the list.

Your understanding is correct, although one thing SA has to deal with (usually by falling on its face) is that the VM could be in an inconsistent state, and there is no safe pointing, locking, or other synchronization it can do to protect from that.

Reply:
Thanks - as per another SA issue my understanding of how the SA operated on a live process was incorrect.